### PR TITLE
Added endpoint to query individuals and get csv as a response

### PR DESCRIPTION
**Pull request by SanjeevLakhwani**
- Now at `/api/individualsgetcsv` one can fetch a post request with a text/json body with key `ids` and value being an array of individual `ids` to get details for those individuals in a csv format with headers

### CODE REVIEW

File: chord_metadata_service/patients/api_views.py [hunk 1]
```diff
@@ -46,6 +49,68 @@ def dispatch(self, *args, **kwargs):
         return super(IndividualViewSet, self).dispatch(*args, **kwargs)
 
 
+class IndividualGetCSVViewSet(viewsets.ModelViewSet):
+    def create(self, request, *args, **kwargs):
```
Comment (ppillot) [inline chord_metadata_service/patients/api_views.py:53]: All this code is duplicated from restapi/api_renderers.py. Would it be possible, within this `create` method to initialize an IndividualCSVRenderer and call its `render` method with the data from the ORM query?

(it appears to me that the query made and the post-processing could be improved in efficiency so a refactor of this code is not out of scope. Please try to limit code duplication here)

**Review by ppillot — CHANGES_REQUESTED** [review]
Thanks for this deep dive in the arcane parts of DRF!
I think that we should have a discussion about the route naming and see if as a group we could come up with a more self-descriptive naming.
Please, also add to this PR the following prerequisites:
- tests: Katsu code base has a very good code coverage. Please add the relevant tests for your new endpoint. You can find examples in the relevant `tests` folder for the object your are writing code for (in your case `chord_metadata_services/patients`). Aim at not lowering the overall code coverage of the application.
- documentation: All API endpoints have been documented. See the /docs folder and add this new endpoint there. Particularly important will be to explain the usage of POST here that is different from the convention used in the rest of the application.

**Comment by ppillot** [comment]
After discussion, here are the requirements:
- rename the endpoint to `batch/individuals`. In the future, other routes starting by `batch` will be implemented
- add a `format` parameter to the body of the POST request. For now, it's acceptable to allow only for CSV format.
- investigate and if possible implement reusing the renderers that were used in the `individuals/` endpoint to avoid code duplication
Thanks!

File: chord_metadata_service/patients/api_views.py [hunk 2]
```diff
@@ -1,3 +1,5 @@
+from django.http import HttpResponse
+import csv
```
[the remark this answers is not in the document]
Comment (SanjeevLakhwani) [reply chord_metadata_service/patients/api_views.py:2]: I have made the changes!

File: chord_metadata_service/restapi/urls.py [hunk 3]
```diff
@@ -29,6 +29,8 @@
 
 # Patients app urls
 router.register(r'individuals', individual_views.IndividualViewSet)
+# router.register(r'individualsgetcsv', individual_views.IndividualViewSetCSV)
```
[the remark this answers is not in the document]
Comment (SanjeevLakhwani) [reply chord_metadata_service/restapi/urls.py:32]: U

File: chord_metadata_service/restapi/urls.py [hunk 4]
```diff
@@ -29,6 +29,8 @@
...
+router.register(r'individualsgetcsv', individual_views.IndividualGetCSVViewSet, basename='individualsgetcsv')
```
[the remark this answers is not in the document]
Comment (SanjeevLakhwani) [reply chord_metadata_service/restapi/urls.py:33]: U

**Review by ppillot — CHANGES_REQUESTED** [review]
Thank-you for the tests and docs addition + route renaming

There were some issues that I felt were left unsolved in this PR, wrt code duplication and also some limitations towards CSV export that were still apparent in some class names.
Instead of returning the PR to you without further guidance due to my lack of knowledge on DRF, I've tried on my side, based on your advances, to reuse the renderers that we already have and leverage what DRF does behind the scenes.

You can find my suggestions as testable code in this branch: https://github.com/bento-platform/katsu/compare/fix/truncated-id-list-on-get?expand=1

Let me present here what I've understood in the process:
- When a request is made, DRF intercepts it with a Router. By default the Router maps GET requests to a ViewSet .details() or .list() methods, POST to .create(), PUT to .update()...
- a ViewSet can define a `queryset` property statically or dynamically using the `get_queryset()` method
- A ViewSet can also define renderers using a list for the `renderer_class` property. A specific renderer is picked via the ContentNegotiation class which by default parses the request to extract the Accept-Type property  and optionally a `format` parameter from the url.
- By default, a ViewSet inherit the .details, .create. list, .update... methods from its base class.
- If I assemble all these pieces, the ViewSet receives the request from the router. It passes its queryset (either static or generated by the get_queryset method) to its Serializer class which generates a result object. This result object is passed to a renderer which has been picked by the ContentNegotiation class, and which returns a Response object.

What I did in the aforementioned branch was to override:
- the router to map POST to .list() and defined no other method
- the ViewSet base class to implement only a .list() method
- the ContentNegotiation class to extract the format parameter from the POST request body and map it to a Renderer or default to the default class behaviour if not found

Finally, the ViewSet could reuse much of the code from the IndividualViewSet (Serializer, Renderers...) and the queryset is defined dynamically using the get_queryset method.

I've encountered issues with the test runner on my machine, which seem due to the usage of POST requests. I had to refactor the tests to each their own class to circumvent this bug.

The last commit of this branch is a proposed change of the parameter from `ids` to `id` as this is a field name and I don't think it should be pluralized, for consistency with the other endpoints. I understand that this is debatable and that using the plural makes is clearer that an array is expected, so feel free to make your choice here.

Please take the necessary time to review the suggestions made and make them your own. I see some advantages: DRY code, avoiding using a `.create()` method to return a content, opening the possibility for multiple export formats (this would need to be tested, as well as the correct rejection of the methods other than POST).
Lastly: if you decide to go for `id` instead of `ids`, the documentation will need to be adapted too.

File: chord_metadata_service/patients/api_views.py [hunk 5]
```diff
@@ -50,6 +55,34 @@ def dispatch(self, *args, **kwargs):
...
+
+    serializer_class = IndividualSerializer
+    pagination_class = LargeResultsSetPagination
+    renderer_classes = (*api_settings.DEFAULT_RENDERER_CLASSES, FHIRRenderer,
+                        PhenopacketsRenderer, IndividualCSVRenderer, ARGORenderer)
+    # Override to infer the renderer based on a `format` argument from the POST request body
+    content_negotiation_class = FormatInPostContentNegotiation
+
+    def get_queryset(self):
+        individual_id = self.request.data.get("id", None)
```
Comment (ppillot) [inline chord_metadata_service/patients/api_views.py:76]: Inconsistent with the latest changes to the tests (id changed back to ids)

**Review by ppillot — CHANGES_REQUESTED** [review]
Github checks are not working.
The change from ids->id is incomplete. The test suite is failing.
Please make sure to run tox locally or at least `manage.py test`.

**Comment by ppillot** [comment]
I finally decided to revert the parameter from `ids` to `id` for consistency with the other endpoints as described in the documentation. I really think we may have more usage of this kind of endpoint in the future and we may introduce a dynamic mapping between POST and GET parameters.

I've updated the documentation to indicate the other types of formats available at this endpoint

When testing with the web service, I've noticed that the pagination of results was effective and I added some code to effectively disable it.

**Pull request merged**